File: test/csharp/SpecGate.CtscFixtures/Conformance/Structured/StructuredOutput.cs
using SpecGate.Annotations;

namespace SpecGate.CtscFixtures.Conformance.Structured;

[SpecEvent("EntityType")]
public sealed class EntityType
{
    [SpecEvent("entity_name")]
    public string Name { get; set; } = string.Empty;

    [SpecEvent("key_properties")]
    public List<string> KeyProperties { get; set; } = [];

    [SpecEvent("structural_properties")]
    public List<string> StructuralProperties { get; set; } = [];
}

public static class StructuredOutput
{
    [SpecOperation("resolve_entity", Spec = "fixture.structured_output")]
    public static EntityType ResolveEntity() =>
        new()
        {
            Name = "Customer",
            KeyProperties = ["ID"],
            StructuralProperties = ["ID", "Name", "Email"],
        };
}

File: test/rust/crates/specgate-ctsc-fixtures/src/conformance/structured/structured_output.rs
use specgate::{SpecEvent, spec_operation};

#[derive(Debug, Clone, PartialEq, Eq, SpecEvent)]
#[spec_component("fixture.structured_output")]
pub struct EntityType {
    #[spec_event(name = "entity_name")]
    pub name: String,
    #[spec_event]
    pub key_properties: Vec<String>,
    #[spec_event]
    pub structural_properties: Vec<String>,
}

#[spec_operation("resolve_entity", spec = "fixture.structured_output")]
pub fn resolve_entity() -> EntityType {
    EntityType {
        name: "Customer".to_string(),
        key_properties: vec!["ID".to_string()],
        structural_properties: vec!["ID".to_string(), "Name".to_string(), "Email".to_string()],
    }
}

#[test]
fn structured_output_projects_renamed_and_list_fields() {
    let entity = resolve_entity();
    assert_eq!(entity.name, "Customer");
    assert_eq!(entity.key_properties, vec!["ID"]);
    assert_eq!(entity.structural_properties.len(), 3);
}
